### ISSUE TITLE
Make fetching all dependencies optional

### ISSUE THREAD

**Issue by flying-sheep** [kingsfordgroup/sailfish#80]
My OS has libbwa and cmph packaged, and I’d rather use those versions than having to read that

`error: downloading 'http://www.cs.stonybrook.edu/~rp/files/bwa-master.tar.gz' failed`

**Comment by roryk** [comment]
+1 for this, libsparsehash fails to build on yosemite which prevents the homebrew-science formula from building: https://github.com/Homebrew/homebrew-science/pull/2706

**Comment by rob-p** [comment]
Hi @roryk, I'm surprised that libsparsehash fails on yosemite, as that's the OSX version I'm working on and the one on which I create the pre-built executable.  In fact, I pull from our own modified branch of [sparsehash](https://github.com/COMBINE-lab/sparsehash) that has fixes specifically aimed at OSX.  Can you tell me what errors, specifically, you're seeing on Yosemite?  For reference, I'm building with Apples provided clang compiler:

```
Apple LLVM version 6.1.0 (clang-602.0.53) (based on LLVM 3.6.0svn)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```
As a side note, this question isn't meant to suggest that I'm not in favor of making dependency fetching optional (I am), but rather I'm also interested in figuring out why you might be experiencing a build failure on a system where I've built successfully before.

**Comment by roryk** [comment]
Thanks @rob-p,

```
Apple LLVM version 7.0.0 (clang-700.0.72)
Target: x86_64-apple-darwin14.5.0
Thread model: posix
```
I wasn't having any troubles with 6.1. 

Here is a gist of the log:
https://gist.github.com/roryk/fda5c55fb732ea06a1b5

And related for the google-sparsehash formula:
https://github.com/Homebrew/homebrew/issues/41087

**Comment by rob-p** [comment]
Ahh --- I haven't upgraded to 7.0.0 yet.  I'll try and take a look at this soon; thanks for alerting me to it!

**Comment by roryk** [comment]
Upgrading Xcode always reminds me of this:

https://www.jwz.org/blog/2012/12/is-there-a-long-german-word-for/

**Comment by flying-sheep** [comment]
all of this doesn’t matter much. i want to be able to install this if alldependencies are there without it trying to download other stuff.

**Comment by rob-p** [comment]
Hi @flying-sheep,

  While I do think that the regression in XCode "matters", you're absolutely right that it's orthogonal the the original issue.  You're also *also* right, of course, that libraries already present on the host OS shouldn't need to be downloaded (unless, like sparsehash, they contain special updates, modifications or fixes that are not present in an upstream version of the package).  I'll try to complete this request this week.

**Comment by roryk** [comment]
Looks like google-sparsehash might have a fix for the 7.0.0 issue:

https://github.com/sparsehash/sparsehash/issues/111

**Comment by rob-p** [comment]
Fantastic!  I was just getting around to working on this.  It actually looks like, after the update (when I pulled out all of the libraries that are no longer used by Sailfish after the update and split with salmon --- e.g. libbwa), there are actually few dependencies whose fetching should be optional.  Right now, the only things we fetch unconditionally are Jellyfish, libgff, and sparsehash.  This is because:
  * We need the dev files for Jellyfish (including the config.h file, since we use it as a library)
  * LibGFF isn't "a thing" --- i.e. it's a small project I created specifically for the purposes of library-izing the gff parser used by Cufflinks, and thus there's no good reason to expect the user would have it pre-installed or to check for it
  * We have to use the version of sparsehash with all of the relevant patches applied to avoid build failure on OSX.

For any other dependencies, we now check if they are installed (can be found via the relevant CMake `findPackage`) before we fetch them.

@roryk --- I've merged the relevant changes to build on Clang 7.0 upstream.  Can you try a clean build and see if it works for you?  You'll have to make sure you've gotten rid of all of CMake's old junk so that it doesn't attempt to use any cached files.

**Comment by rob-p** [comment]
Of course, let me [un-botch the Travis-CI build](https://travis-ci.org/kingsfordgroup/sailfish) first ... Sigh.

**Comment by rob-p** [comment]
Ok --- fixed (there were some small changes with the latest jellyfish that I had to adjust for).

**Comment by roryk** [comment]
Thanks @rob-p, builds fine and the test passes under 7.0.0 now. Sorry to hijack this issue.

**Comment by roryk** [comment]
Could you push a new release so I can fix the homebrew-science formula? :)

**Comment by rob-p** [comment]
[Done](https://github.com/kingsfordgroup/sailfish/releases/tag/v0.7.5) :).

**Comment by sjackman** [comment]
:+1: I've updated the Homebrew-science formula to 0.9.2.

**Comment by mmokrejs** [comment]
Well I am trying to move ahead with the packe for Gentoo Linux so it does not really help. Bundled/autodownloaded approach could remain for users who could even download a bunch of binaries and libs but the source tarball should be just clean.
